### PR TITLE
B3 Groups + related fixes

### PR DESCRIPTION
**Pull request by biyeun**
- reverts reverted commit https://github.com/dimagi/commcare-hq/commit/198a6b6252bda79833ac4434b4c70e2021d2b193
- fixes issue with reset password form in user details page
- makes sure other forms on page submit properly
- adds new multiselect widget in groups and locations
  screen shot 2016-05-25 at 5 09 53 pm
  screen shot 2016-05-25 at 5 09 44 pm
  screen shot 2016-05-25 at 5 09 27 pm
  screen shot 2016-05-25 at 5 09 12 pm
  screen shot 2016-05-25 at 3 18 21 pm

@benrudolph @orangejenny cc: @czue & @dimagi/product

### CODE REVIEW

**Comment by dimagimon** [comment]
There are 13 errors:

* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 17 - 'django' is not defined. (no-undef)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 40 - Expected indentation of 20 space characters but found 18. (indent)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 83 - Missing trailing comma. (comma-dangle)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 41 - Expected indentation of 22 space characters but found 20. (indent)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 31 - 'ms' is defined but never used (no-unused-vars)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 42 - Expected indentation of 22 space characters but found 20. (indent)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 86 - 'e' is defined but never used (no-unused-vars)
* corehq/apps/users/views/__init__.py, line 65 - F401 'UpdateMyAccountInfoForm' imported but unused
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 56 - Expected indentation of 20 space characters but found 18. (indent)
Expected '===' and instead saw '=='. (eqeqeq)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 57 - Expected indentation of 22 space characters but found 20. (indent)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 25 - 'django' is not defined. (no-undef)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 90 - 'e' is defined but never used (no-unused-vars)
* corehq/apps/style/static/style/js/components/multiselect_utils.js, line 58 - Expected indentation of 22 space characters but found 20. (indent)

File: corehq/apps/style/static/style/js/components/multiselect_utils.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        selectableHeaderTitle,
+        selectedHeaderTitle,
+        searchItemTitle
+    ) {
+        var selectAllId = multiselectId + '-select-all',
+            removeAllId = multiselectId + '-remove-all',
+            searchSelectableId = multiselectId + '-search-selectable',
+            searchSelectedId = multiselectId + '-search-selected';
+        $('#' + multiselectId).multiSelect({
+            selectableHeader: '<div class="ms-header">' + selectableHeaderTitle +
```
Comment (benrudolph) [inline corehq/apps/style/static/style/js/components/multiselect_utils.js]: any chance we can use underscore templates for these?

File: corehq/apps/style/static/style/js/components/multiselect_utils.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                    if (that.qs1.val().length > 0) {
+                        $('#' + selectAllId).addClass('disabled').prop('disabled', true);
+                    } else {
+                        $('#' + selectAllId).removeClass('disabled').removeProp('disabled');
+                    }
+                });
+
+                that.qs2 = $selectionSearch.quicksearch(selectionSearchString)
+                .on('keydown', function () {
+                    if (e.which === 40) {
```
Comment (dimagimon) [inline corehq/apps/style/static/style/js/components/multiselect_utils.js]: 'e' is not defined. (no-undef)

File: corehq/apps/style/static/style/js/components/multiselect_utils.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                this.qs1.cache();
+                this.qs2.cache();
+            },
+        });
+
+        $('#' + selectAllId).click(function(e){
+            $('#' + multiselectId).multiSelect('select_all');
+            return false;
+        });
+        $('#' + removeAllId).click(function(e){
```
Comment (dimagimon) [inline corehq/apps/style/static/style/js/components/multiselect_utils.js]: 'e' is defined but never used (no-unused-vars)

File: corehq/apps/style/static/style/js/components/multiselect_utils.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            afterInit: function () {
+                var that = this,
+                    $selectableSearch = $('#'+searchSelectableId),
+                    $selectionSearch = $('#'+searchSelectedId),
+                    selectableSearchString = '#'+that.$container.attr('id')+' .ms-elem-selectable:not(.ms-selected)',
+                    selectionSearchString = '#'+that.$container.attr('id')+' .ms-elem-selection.ms-selected';
+
+                that.qs1 = $selectableSearch.quicksearch(selectableSearchString)
+                .on('keydown', function () {
+                    if (e.which === 40) {
```
Comment (dimagimon) [inline corehq/apps/style/static/style/js/components/multiselect_utils.js]: 'e' is not defined. (no-undef)

File: corehq/apps/style/static/style/js/components/multiselect_utils.js [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            afterDeselect: function(){
+                // remove search option so that user doesn't get confused
+                this.qs1.val('').search('');
+                $('#' + selectAllId).removeClass('disabled').removeProp('disabled');
+                this.qs1.cache();
+                this.qs2.cache();
+            },
+        });
+
+        $('#' + selectAllId).click(function(e){
```
Comment (dimagimon) [inline corehq/apps/style/static/style/js/components/multiselect_utils.js]: 'e' is defined but never used (no-unused-vars)

**Comment by benrudolph** [comment]
so much 😂 !! 💀 to 
screen shot 2016-05-25 at 5 22 07 pm

**Comment by amsagoff** [comment]
end of an era

File: corehq/apps/style/static/style/js/components/multiselect_utils.js [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+            '<a href="#" class="btn <%=actionButtonClass %> btn-xs pull-right" id="<%= actionButtonId %>">' +
+                '<i class="<%= actionButtonIcon %>"></i> <%= actionButtonText %>' +
+            '</a>'
+        );
+        return action({
+            actionButtonId: buttonId,
+            actionButtonClass: buttonClass,
+            actionButtonIcon: buttonIcon,
+            actionButtonText: text,
+        })
```
Comment (dimagimon) [inline corehq/apps/style/static/style/js/components/multiselect_utils.js]: Missing semicolon. (semi)

**Comment by biyeun** [comment]
hold off on merge while we hallway test on staging

File: corehq/apps/style/static/style/js/components/multiselect_utils.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+            afterInit: function(ms){
+                var that = this,
+                    $selectableSearch = $('#'+searchSelectableId),
+                    $selectionSearch = $('#'+searchSelectedId),
+                    selectableSearchString = '#'+that.$container.attr('id')+' .ms-elem-selectable:not(.ms-selected)',
+                    selectionSearchString = '#'+that.$container.attr('id')+' .ms-elem-selection.ms-selected';
+
+                that.qs1 = $selectableSearch.quicksearch(selectableSearchString)
+                .on('keydown', function(e){
+                  if (e.which === 40){
```
Comment (orangejenny) [inline corehq/apps/style/static/style/js/components/multiselect_utils.js]: comment what key code this is (same below)

File: corehq/apps/style/static/style/js/components/multiselect_utils.js [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+                                    '<input type="search" class="form-control search-input" id="' + searchSelectedId + '" autocomplete="off" placeholder="' + searchItemTitle + '" />' +
+                               '</div>',
+            afterInit: function(ms){
+                var that = this,
+                    $selectableSearch = $('#'+searchSelectableId),
+                    $selectionSearch = $('#'+searchSelectedId),
+                    selectableSearchString = '#'+that.$container.attr('id')+' .ms-elem-selectable:not(.ms-selected)',
+                    selectionSearchString = '#'+that.$container.attr('id')+' .ms-elem-selection.ms-selected';
+
+                that.qs1 = $selectableSearch.quicksearch(selectableSearchString)
```
Comment (orangejenny) [inline corehq/apps/style/static/style/js/components/multiselect_utils.js]: Nitpick, but `qs1` and `qs1` are vague. Maybe `search_left` and `search_right`?

**Comment by orangejenny** [comment]
this looks awesome!

**Comment by orangejenny** [comment]
code looks good, just a couple non-blocking comments

File: corehq/apps/style/static/style/js/components/multiselect_utils.js (hunk 8 above)
Comment (biyeun) [inline corehq/apps/style/static/style/js/components/multiselect_utils.js]: literally copy pasted as the solution from the loudev docs

File: custom/ilsgateway/templates/ilsgateway/slab/edit_location.html [hunk 9]
```diff
@@ -3,19 +3,20 @@
...
 {% endblock %}
 
 {% block js-inline %}{{ block.super }}
     <script type="text/javascript">
         $(function () {
-            $("#id_selected_ids").width(800).height(400).multiselect();
+            var multiselect_utils = hqImport('style/js/components/multiselect_utils');
+            multiselect_utils.createFullMultiselectWidget(
+                'id_selected_ids',
+                django.gettext("Available Locations"),
```
Comment (biyeun) [inline custom/ilsgateway/templates/ilsgateway/slab/edit_location.html]: @czue @kkrampa does this text make sense in the context of this widget? I'm not familiar with this UI to say 100% that this is a good word choice.

Comment (czue) [inline custom/ilsgateway/templates/ilsgateway/slab/edit_location.html]: I believe so, but @kkrampa or @modonnell729 should confirm

**Comment by czue** [comment]
this looks awesome!

**Comment by biyeun** [comment]
fyi, resolved the merge conflicts so waiting on ✅

**Comment by biyeun** [comment]
@orangejenny / @benrudolph cc @calellowitz ✅

File: custom/ilsgateway/templates/ilsgateway/slab/edit_location.html (hunk 9 above)
Comment (biyeun) [inline custom/ilsgateway/templates/ilsgateway/slab/edit_location.html]: @kkrampa @modonnell729 any thoughts? regardless, this shouldn't block merge

**Comment by biyeun** [comment]
also @czue ✅

**Pull request merged**